Bottom: 89a454ee98f3470b30fa9aa778f567750f93a8a0
Top:    5ea685cdd66a1eb2652ee9d6ead781c8cc4a6a5d
Author: Robert Collins <robertc@robertcollins.net>
Date:   2013-03-06 00:01:37 +1300

Refresh of placeholder-tags

---

diff --git a/NEWS b/NEWS
index 78013ee..cd9a370 100644
--- a/NEWS
+++ b/NEWS
@@ -23,6 +23,12 @@ Improvements
 * The error message for setUp and tearDown upcall errors was broken on Python
   3.4. (Monty Taylor, Robert Collins, #1140688)
 
+Improvements
+------------
+
+* ``PlaceHolder`` can now hold tags, and applies them before, and removes them
+  after, the test. (Robert Collins)
+
 0.9.29
 ~~~~~~
 
diff --git a/testtools/testcase.py b/testtools/testcase.py
index 3998ace..97158c0 100644
--- a/testtools/testcase.py
+++ b/testtools/testcase.py
@@ -618,7 +618,7 @@ class PlaceHolder(object):
     failureException = None
 
     def __init__(self, test_id, short_description=None, details=None,
-        outcome='addSuccess', error=None):
+        outcome='addSuccess', error=None, tags=None):
         """Construct a `PlaceHolder`.
 
         :param test_id: The id of the placeholder test.
@@ -626,6 +626,7 @@ class PlaceHolder(object):
             test. If not provided, the id will be used instead.
         :param details: Outcome details as accepted by addSuccess etc.
         :param outcome: The outcome to call. Defaults to 'addSuccess'.
+        :param tags: Tags to report for the test.
         """
         self._test_id = test_id
         self._short_description = short_description
@@ -633,6 +634,7 @@ class PlaceHolder(object):
         self._outcome = outcome
         if error is not None:
             self._details['traceback'] = content.TracebackContent(error, self)
+        self._tags = tags or set()
 
     def __call__(self, result=None):
         return self.run(result=result)
@@ -666,10 +668,12 @@ class PlaceHolder(object):
 
     def run(self, result=None):
         result = self._result(result)
+        result.tags(self._tags, set())
         result.startTest(self)
         outcome = getattr(result, self._outcome)
         outcome(self, details=self._details)
         result.stopTest(self)
+        result.tags(set(), self._tags)
 
     def shortDescription(self):
         if self._short_description is None:
diff --git a/testtools/tests/test_testcase.py b/testtools/tests/test_testcase.py
index 127f8d0..915aa47 100644
--- a/testtools/tests/test_testcase.py
+++ b/testtools/tests/test_testcase.py
@@ -117,7 +117,8 @@ class TestPlaceHolder(TestCase):
         log = []
         test.run(LoggingResult(log))
         self.assertEqual(
-            [('startTest', test), ('addSuccess', test), ('stopTest', test)],
+            [('tags', set(), set()), ('startTest', test), ('addSuccess', test),
+             ('stopTest', test), ('tags', set(), set()),],
             log)
 
     def test_supplies_details(self):
@@ -126,9 +127,12 @@ class TestPlaceHolder(TestCase):
         result = ExtendedTestResult()
         test.run(result)
         self.assertEqual(
-            [('startTest', test),
+            [('tags', set(), set()),
+             ('startTest', test),
              ('addSuccess', test, details),
-             ('stopTest', test)],
+             ('stopTest', test),
+             ('tags', set(), set()),
+             ],
             result._events)
 
     def test_call_is_run(self):
@@ -149,6 +153,19 @@ class TestPlaceHolder(TestCase):
         # A PlaceHolder can be debugged.
         self.makePlaceHolder().debug()
 
+    def test_supports_tags(self):
+        result = ExtendedTestResult()
+        tags = set(['foo', 'bar'])
+        case = PlaceHolder("foo", tags=tags)
+        case.run(result)
+        self.assertEqual([
+            ('tags', tags, set()),
+            ('startTest', case),
+            ('addSuccess', case),
+            ('stopTest', case),
+            ('tags', set(), tags),
+            ], result._events)
+
 
 class TestErrorHolder(TestCase):
     # Note that these tests exist because ErrorHolder exists - it could be
@@ -202,9 +219,11 @@ class TestErrorHolder(TestCase):
         log = result._events
         test.run(result)
         self.assertEqual(
-            [('startTest', test),
+            [('tags', set(), set()),
+             ('startTest', test),
              ('addError', test, test._details),
-             ('stopTest', test)], log)
+             ('stopTest', test),
+             ('tags', set(), set())], log)
 
     def test_call_is_run(self):
         # A PlaceHolder can be called, in which case it behaves like run.
